ocaml/xapi/xapi_clustering.ml

```diff
@@ -241,6 +221,46 @@ let rpc ~__context =
   | None ->
     Cluster_client.rpc (fun () -> failwith "Can only communicate with xapi-clusterd through message-switch")
 
+let assert_cluster_host_quorate ~__context ~self =
+  (* With the latest kernel GFS2 would hang on mount if clustering is not working yet,
+   * whereas previously we got a 'Transport endpoint not connected' error.
+   * Ensure that we are quorate now: even if we have enabled the cluster host we may not have
+   * achieved quorum yet if we have just booted and haven't seen enough hosts.
+   * Do this via an API call rather than reading a field in the database, because the field in the
+   * database could be out of date.
+   * *)
+  let result = Cluster_client.LocalClient.diagnostics (rpc ~__context) "assert_cluster_host_quorate"
+  in
+  match Idl.IdM.run @@ Cluster_client.IDL.T.get result with
+  | Result.Ok diag ->
+    debug "Local cluster host is quorate: %b" diag.Cluster_interface.is_quorate;
+    if not diag.Cluster_interface.is_quorate then
+        raise Api_errors.(Server_error (cluster_host_not_joined, [Ref.string_of self]))
+  | Result.Error error ->
+    warn "Cannot query cluster host quorate status";
+    handle_error error
+
+let assert_cluster_host_is_enabled_for_matching_sms ~__context ~host ~sr_sm_type =
+  begin match get_required_cluster_stacks ~__context ~sr_sm_type with
+    | [] -> ()
+    | required_cluster_stacks ->
+      (* One of these [required_cluster_stacks] should be configured and running *)
+      let cluster_stack_of ~cluster_host =
+        let cluster = Db.Cluster_host.get_cluster ~__context ~self:cluster_host in
+        Db.Cluster.get_cluster_stack ~__context ~self:cluster
+      in
+      let error_no_cluster_host_found condition =
+        debug "No_cluster_host found%s" condition;
+        raise Api_errors.(Server_error (no_compatible_cluster_host, [Ref.string_of host]))
+      in
+      match find_cluster_host ~__context ~host with
+        | Some cluster_host when (List.mem (cluster_stack_of ~cluster_host) required_cluster_stacks) ->
```

Comment:
The indentation for `match` looks different from above. Maybe run this through xcp-indent

Reply:
Done, I think this code got moved around quite a bit.